### ISSUE TITLE
portal.sinric.pro Thermostat mode setting not working

### ISSUE THREAD

**Issue by marcosmoura91** [sinricpro/esp8266-esp32-sdk#91]
Hello everyone,

I'm not sure if here is the right place to post this but using the Portal I cannot update my thermostat mode.
I am using debuging mode and I get no message when i set the mode, Cooling, Auto, Heat all aren't working.

For turning off, on and set point all work. I tested using Portal and also using alexa echo. The request to change the mode is not being sent from your server to my device.

**Comment by sivar2311** [comment]
I did a check, and everything worked fine!
Serial log (IP and DeviceId is x`ed out):
```
[Wifi]: Connecting.............connected!
[WiFi]: IP-Address is xxx.xxx.xxx.xxx       
Connected to SinricPro
Thermostat 5fxxxxxxxxxxxxxxxxxxxxxxxxx set to mode COOL
Thermostat 5fxxxxxxxxxxxxxxxxxxxxxxxxx set to mode HEAT
Thermostat 5fxxxxxxxxxxxxxxxxxxxxxxxxx set to mode AUTO
```
Using the given [example](https://github.com/sinricpro/esp8266-esp32-sdk/blob/master/examples/Thermostat/Thermostat.ino)

**Comment by marcosmoura91** [comment]
Hi, so I copy paste your example and only added my personal settings and still I can do everything except for the AC mode.

What I just noticed is that it says my device doesn't exist but I can see it on the Portal and I copy paste all ID numbers i double checked for mistakes and all good.

[SinricPro]: Device "XXXXXXXXXX" does not exist. Creating new device.

> [Wifi]: Connecting.............connected!
> [WiFi]: IP-Address is 192.168.1.24
> Connected to SinricPro
> Thermostat 5e4b3945aeb6075deb11e26d turned on
> 

Now this is my debug: 

I am using the portal to click on turn on / off , set temperature and mode. When I set a mode I get no message no feedback on my esp.

> [Wifi]: Connecting.................connected!
> [WiFi]: IP-Address is 192.168.1.24
> [SinricPro]: Device "XXXXXXXXXX" does not exist. Creating new device
> [SinricPro:add()]: Adding device with id "XXXXXXXX".
> [SinricPro:Websocket]: Connecting to WebSocket Server (ws.sinric.pro)
> [SinricPro:Websocket]: headers: 
> appkey:XXXXXXXX
> deviceids:XXXXXXXXX
> restoredevicestates:false
> ip:192.168.1.24
> mac:CC:50:E3:17:XX:XX
> platform:ESP8266
> version:2.7.0
> [SinricPro:Websocket]: connected
> Connected to SinricPro
> [SinricPro:Websocket]: receiving data
> [SinricPro.handleReceiveQueue()]: 1 message(s) in receiveQueue
> [SinricPro.handleReceiveQueue()]: Signature is valid. Processing message...
> [SinricPro:extractTimestamp(): Got Timestamp 1602025930
> [SinricPro:Websocket]: receiving data
> [SinricPro.handleReceiveQueue()]: 1 message(s) in receiveQueue
> [SinricPro.handleReceiveQueue()]: Signature is valid. Processing message...
> [SinricPro:extractTimestamp(): Got Timestamp 1602026022
> [SinricPro.handleRequest()]: handling request
> {
>   "header": {
>     "payloadVersion": 2,
>     "signatureVersion": 1
>   },
>   "payload": {
>     "action": "setPowerState",
>     "clientId": "sinricpro-web",
>     "createdAt": 1602026022,
>     "deviceAttributes": [],
>     "deviceId": "XXXXX",
>     "replyToken": "4b942677-510c-4ba5-ae93-36e3XXXXXX",
>     "type": "request",
>     "value": {
>       "state": "On"
>     }
>   },
>   "signature": {
>     "HMAC": "iiXXXXrriN/w4wzUYqXXXXXSFBZiASBFSKv/ew="
>   }
> }SinricProDevice::handleRequest()
> Thermostat XXXXXXX turned on
> [SinricPro:handleSendQueue()]: 1 message(s) in sendQueue
> [SinricPro:handleSendQueue()]: Sending message...
> {
>   "header": {
>     "payloadVersion": 2,
>     "signatureVersion": 1
>   },
>   "payload": {
>     "action": "setPowerState",
>     "clientId": "sinricpro-web",
>     "createdAt": 1602026023,
>     "deviceId": "XXXXXX",
>     "message": "OK",
>     "replyToken": "4b9XXXXX-510c-4ba5-ae93-36XXXX0627",
>     "success": true,
>     "type": "response",
>     "value": {
>       "state": "On"
>     }
>   },
>   "signature": {
>     "HMAC": "ipVBMXXXX2a1uYcZXXXXXXMcvXczYXXXXX8bstr7CY="
>   }
> }
> [SinricPro:handleSendQueue]: Sending to websocket
> [SinricPro:handleSendQueue()]: message sent.
> 



Here is my code: 

```
// Uncomment the following line to enable serial debug output
#define ENABLE_DEBUG

#ifdef ENABLE_DEBUG
       #define DEBUG_ESP_PORT Serial
       #define NODEBUG_WEBSOCKETS
       #define NDEBUG
#endif 

#include <Arduino.h>
#ifdef ESP8266 
       #include <ESP8266WiFi.h>
#endif 
#ifdef ESP32   
       #include <WiFi.h>
#endif

#include "SinricPro.h"
#include "SinricProThermostat.h"

#define WIFI_SSID         "xxxx"    
#define WIFI_PASS         "xxxx"
#define APP_KEY           "5ecef3xxxxxxx"      // Should look like "de0ax2abxxxxxxxx"
#define APP_SECRET        "axxxxxx"   // Should look like "5f36xxxx-x3x7-4x3x-xexe-e86724a9xxxx-4c4axxxx-3x3x-x5xe-x9x3-333d65xxxxxx"
#define THERMOSTAT_ID     "5e4bxxxxxx"    // Should look like "5dc1564130xxxxxx"
#define BAUD_RATE         9600                     // Change baudrate to your need

float globalTemperature;
bool globalPowerState;

bool onPowerState(const String &deviceId, bool &state) {
  Serial.printf("Thermostat %s turned %s\r\n", deviceId.c_str(), state?"on":"off");
  globalPowerState = state; 
  return true; // request handled properly
}

bool onTargetTemperature(const String &deviceId, float &temperature) {
  Serial.printf("Thermostat %s set temperature to %f\r\n", deviceId.c_str(), temperature);
  globalTemperature = temperature;
  return true;
}

bool onAdjustTargetTemperature(const String & deviceId, float &temperatureDelta) {
  globalTemperature += temperatureDelta;  // calculate absolut temperature
  Serial.printf("Thermostat %s changed temperature about %f to %f", deviceId.c_str(), temperatureDelta, globalTemperature);
  temperatureDelta = globalTemperature; // return absolut temperature
  return true;
}

bool onThermostatMode(const String &deviceId, String &mode) {
  Serial.printf("Thermostat %s set to mode %s\r\n", deviceId.c_str(), mode.c_str());
  return true;
}


void setupWiFi() {
  Serial.printf("\r\n[Wifi]: Connecting");
  WiFi.begin(WIFI_SSID, WIFI_PASS);

  while (WiFi.status() != WL_CONNECTED) {
    Serial.printf(".");
    delay(250);
  }
  IPAddress localIP = WiFi.localIP();
  Serial.printf("connected!\r\n[WiFi]: IP-Address is %d.%d.%d.%d\r\n", localIP[0], localIP[1], localIP[2], localIP[3]);
}

void setupSinricPro() {
  SinricProThermostat &myThermostat = SinricPro[THERMOSTAT_ID];
  myThermostat.onPowerState(onPowerState);
  myThermostat.onTargetTemperature(onTargetTemperature);
  myThermostat.onAdjustTargetTemperature(onAdjustTargetTemperature);
  myThermostat.onThermostatMode(onThermostatMode);

  // setup SinricPro
  SinricPro.onConnected([](){ Serial.printf("Connected to SinricPro\r\n"); }); 
  SinricPro.onDisconnected([](){ Serial.printf("Disconnected from SinricPro\r\n"); });
  SinricPro.begin(APP_KEY, APP_SECRET);
}

void setup() {
  Serial.begin(BAUD_RATE); Serial.printf("\r\n\r\n");
  setupWiFi();
  setupSinricPro();
}

void loop() {
  SinricPro.handle();
}

```

**Comment by marcosmoura91** [comment]
So I just did a 2nd test and used a thermostat I set on my beedroom room instead of my office room (as per sinric settings). Same code just device id is different.

I still get the message [SinricPro]: Device "5e4bxxxx" does not exist. Creating new device

but now the AC mode works now.
I reverted the device id back to the one setup on my office room and still no AC mode message comming in.

Note that this test is with the same hardware, same device. just changing the deviceid.

I set my Office_AC to be attached to the bedroom room on the sinric website and still doesn't work.
I completly deleted the device and added a new thermostat to the bedroom which now I have 2. Office_AC and bedroom_AC and also doesn't work.

Is there a problem by having multiple thermostats where only the first one will work ? I am very confused

**Comment by sivar2311** [comment]
> I still get the message [SinricPro]: Device "5e4bxxxx" does not exist. Creating new device

This is okay. Its a message from SDK that means that a device have been created (inside the SDK / your ESP). 

> but now the AC mode works now.
> I reverted the device id back to the one setup on my office room and still no AC mode message comming in.
> 
> Note that this test is with the same hardware, same device. just changing the deviceid.

I think there might be a problem with the deviceId.
Easy solution is to create a new device (like you did) and delete the old one. 
But for more investigation we need the (faulty) deviceId.
 
> I set my Office_AC to be attached to the bedroom room on the sinric website and still doesn't work.
> I completly deleted the device and added a new thermostat to the bedroom which now I have 2. Office_AC and bedroom_AC and also doesn't work.
> 

Now i am confused. Was this related to the first (faulty) device or the new device?

> **At the end using ALEXA it's working now. I can set the mode by speaking to it. But setting the mode on the website sinric for my 2nd AC does not work.**
> 

Generally, if alexa voice commands work, portal should work too.
Another problem might be the name. If they sounds too similar, alexa could get confused and may use the wrong device.

**Comment by marcosmoura91** [comment]
Hi, I deleted the faulty device and added a new one to my bedroom and tried using the new device and it didnt work also.
Only the first thermostat works using the website GUI. Any new thermostat that I add only works with Alexa but not using the GUI. 
Did you try adding a new thermostat and tried setting the mode using the website GUI, did it work ?

**Comment by sivar2311** [comment]
Can you provide a current screenshot please?
I think the other thermostat is not online. See red (offline) or green (online) line above the device!
Since the names are very similar, alexa is using the one which is online!

**Comment by marcosmoura91** [comment]
Hi, I only have once device at the moment, in the arduino code I just switch between device IDs so there will be always one disconnected. My problem is, when I add a 2nd thermostat device and configure it on my esp8266 if I try to change the mode using the GUI it doesn't work but everything else works, On, Off and Temperature setting. 

If I change my code on my ESP8266 to the first device again (AC) then the Mode setting works again

**Comment by sivar2311** [comment]
For testing: Please let both thermostats run simultaneously on the ESP.

**Comment by marcosmoura91** [comment]
Hi, ok so I just tested both online at the same time it works i receive mode change request for each individually.
But If I have the first one offline and use only the second one set at my office room then it doesn't work..

I will post here my code: 

```
#define ENABLE_DEBUG

#ifdef ENABLE_DEBUG
       #define DEBUG_ESP_PORT Serial
       #define NODEBUG_WEBSOCKETS
       #define NDEBUG
#endif 

#include <Arduino.h>
#ifdef ESP8266 
       #include <ESP8266WiFi.h>
#endif 
#ifdef ESP32   
       #include <WiFi.h>
#endif

#include "SinricPro.h"
#include "SinricProThermostat.h"

#define WIFI_SSID         "xxxx"    
#define WIFI_PASS         "xxxx"
#define APP_KEY           "5ecef3xxxx"      // Should look like "de0bxxxx-1x3x-4x3x-ax2x-5dabxxxxxxxx"                      
#define APP_SECRET        "aefda0xxxxx"   // Should look like "5f36xxxx-x3x3-333d65xxxxxx"                         
#define THERMOSTAT_ID     "5f7d007d2xxxx"    // Work AC
//#define THERMOSTAT_ID2    "5f7d007d20yyyy"    // Bedroom AC
#define BAUD_RATE         9600                     // Change baudrate to your need

float globalTemperature;
bool globalPowerState;

bool onPowerState(const String &deviceId, bool &state) {
  Serial.printf("Thermostat %s turned %s\r\n", deviceId.c_str(), state?"on":"off");
  globalPowerState = state; 
  return true; // request handled properly
}

bool onTargetTemperature(const String &deviceId, float &temperature) {
  Serial.printf("Thermostat %s set temperature to %f\r\n", deviceId.c_str(), temperature);
  globalTemperature = temperature;
  return true;
}

bool onAdjustTargetTemperature(const String & deviceId, float &temperatureDelta) {
  globalTemperature += temperatureDelta;  // calculate absolut temperature
  Serial.printf("Thermostat %s changed temperature about %f to %f", deviceId.c_str(), temperatureDelta, globalTemperature);
  temperatureDelta = globalTemperature; // return absolut temperature
  return true;
}

bool onThermostatMode(const String &deviceId, String &mode) {
  Serial.printf("Thermostat %s set to mode %s\r\n", deviceId.c_str(), mode.c_str());
  return true;
}


void setupWiFi() {
  Serial.printf("\r\n[Wifi]: Connecting");
  WiFi.begin(WIFI_SSID, WIFI_PASS);

  while (WiFi.status() != WL_CONNECTED) {
    Serial.printf(".");
    delay(250);
  }
  IPAddress localIP = WiFi.localIP();
  Serial.printf("connected!\r\n[WiFi]: IP-Address is %d.%d.%d.%d\r\n", localIP[0], localIP[1], localIP[2], localIP[3]);
}

void setupSinricPro() {
  SinricProThermostat &myThermostat = SinricPro[THERMOSTAT_ID];
  //SinricProThermostat &myThermostat2 = SinricPro[THERMOSTAT_ID2];
  myThermostat.onPowerState(onPowerState);
  myThermostat.onTargetTemperature(onTargetTemperature);
  myThermostat.onAdjustTargetTemperature(onAdjustTargetTemperature);
  myThermostat.onThermostatMode(onThermostatMode);

  //myThermostat2.onPowerState(onPowerState);
  //myThermostat2.onTargetTemperature(onTargetTemperature);
  //myThermostat2.onAdjustTargetTemperature(onAdjustTargetTemperature);
  //myThermostat2.onThermostatMode(onThermostatMode);

  // setup SinricPro
  SinricPro.onConnected([](){ Serial.printf("Connected to SinricPro\r\n"); }); 
  SinricPro.onDisconnected([](){ Serial.printf("Disconnected from SinricPro\r\n"); });
  SinricPro.begin(APP_KEY, APP_SECRET);
}

void setup() {
  Serial.begin(BAUD_RATE); Serial.printf("\r\n\r\n");
  setupWiFi();
  setupSinricPro();
}

void loop() {
  SinricPro.handle();
}
```

**Comment by sivar2311** [comment]
> By uncommenting the 2nd Thermostat then both work using the GUI.

This is confusing! 
When there is just one thermostat online, how both can work??
Please double check your deviceIds.

What do you want to acheive? Running two devices, or just one?

**Comment by marcosmoura91** [comment]
I explain you the problem. I just try to recreate it using one ESP8266.

I have one ESP8266 on my bedroom controlling my AC the name is AC as you see on the picture.
I have a second ESP8266 on my Office room named work AC. 

If my Bedroom AC esp8266 is OFF, then as consequence, I cannot control my second ESP8266 using the website GUI.

**Comment by sivar2311** [comment]
I'cant imagine this.
Ok, please delete both devices in portal and create new ones using new deviceId numbers.
Update your two sketches to the new ID's, upload and connect them to SinricPro.
In Portal, both devices should shown as online.

**Comment by marcosmoura91** [comment]
Doesn't work. 

I just deleted both Thermostats. Added two new ones. One to the Bedroom and one to my Office room. I even changed the names one is "bed AC" Another is "Offic AC" both in different rooms as the picture above.

Same problem happens. I can only set the mode for my Offic AC if I delete the bed room thermostat. The only way my Office Thermostat GUI mode setting works is only if: 

- Bedroom thermostat is On
- Deleting my bedroom thermostat.

Very strange I know...  Somehow the thermostat from the first room effects the others... Do I have to delete all my rooms and add them again ??

**Comment by sivar2311** [comment]
Please upload both sketches to gist.github.com and post the link here. 
I will download and try to reproduce using two NodeMCUs.

**Comment by marcosmoura91** [comment]
I am using the Thermostat Example. No changes. Just added my Api ID and device ID. The same example you sent me for testing just copy and paste.

**Comment by sivar2311** [comment]
Please send your sketches via e-mail to me (including your deviceIds and a screenshot of your devices tab in portal)
sivar2311@gmail.com

**Comment by sivar2311** [comment]
btw: Did you check "Activity Log" in portal?

**Comment by marcosmoura91** [comment]
Ok so I checked the activity log, 

First for double check I sent for each an ON command. and you can see position 3 and 4 of the log.
After this I set the mode for the first and the for the 2nd thermostat and as you can see both request go to the same Thermostat ! Some how the GUI is sending the 2nd request to the wrong thermostat.

https://ibb.co/94QyRzm

**Comment by sivar2311** [comment]
I was able to reproduce the issue.
@kakopappa If two thermostat devices are running, all setThermostatMode-commands are sent to only one device

**Comment by kakopappa** [comment]
Thanks everyone!

This will be fixed and shipped with Portal v2.11.1. Fix will be in the production with-in a week's time.

**Comment by kakopappa** [comment]
Hi everyone. this issue fix is in the production now. (v2.10.2)
